### DISCUSSION TITLE
walk() should have a non-None per-default stop_type depending on certain values of my_type

### DISCUSSION THREAD

**Discussion by schreiberx** [stfc/PSyclone#2794] [Ideas]
Without testing this explicitly, I assume it will improve the performance of walking over psyir trees significantly in many cases if setting a 2nd default `stop_type` (in case that `stop_type == None` if `my_type` is set to a type that can't be used recursively.

E.g., `Call`, `Container` (the pure container, not FileContainer) and `Routine` can't be used recursively (in Fortran).
Therefore, I suggest setting the `stop_type` per default for a certain values of `my_type` (`Routine`, `Call`, `Container`) simply to the value of `my_type`.

@arporter @sergisiso

**Comment by sergisiso** [comment]
Thanks for the suggestion @schreiberx, I see what you mean, but remember that the PSyIR is a canonicalized version of the Fortran semantics (and this is good to make the transformation scripts easy and broadly applicable). Therefore these concepts that you mention can be recursive in the PSyIR, for example Call includes calls to routine, functions and intrinsics, so:
`call my_sub(my_func(max(3,4)))` is represented by 3 nested PSyIR Call nodes.

Similarly for Container: it is used for the file scope which can have multiple modules inside, and also subroutines can have their own "contains" sections inside (currently creates a Codeblock but some day may mean nested containers). And for subroutines: nested routines are possible in Fortran (as mentioned before), and in the future we hay want to support statements functions as a nested routine.

In addition the time of walks are barely visible in the profiler now, so I would avoid doing this optimisation even if we really find and edge case where this is possible.

**Comment by schreiberx** [comment]
Indeed, Calls shouldn't belong on the list of the suggested cyclone nodes.
And for the Routine, I now see that Fortran has "more" features than I'd have expected :-)

I'm adding this link to the nested "routines" and will close this discussion.
https://craftofcoding.wordpress.com/2021/01/15/coding-fortran-nested-functions/

Thanks for your answer!
